### DISCUSSION TITLE
`pnpm projen` seems to revert `pnpm upgrade`

### DISCUSSION THREAD

**Discussion by bestickley** [projen/projen#3905] [Q&A] [answered]
When I run `pnpm upgrade` in [cdk-nextjs](https://github.com/cdklabs/cdk-nextjs) everything looks great as my dependencies are being updated following semver. So I commit and push to make a PR, but then I get a frozen lockfile error in GitHub Action. So then I run `pnpm projen` locally and I see that it reverts all the updates in my package.json and reverts the lockfile. Is this expected? Could someone explain this to me? Thank you!

UPDATE:
so I tried this again and it actually worked. I think before I ran `pnpm update` (not upgrade) which updates core deps to project like jsii and others. When I run `pnpm projen` after that they're reverted because like @mrgrain says below, they're managed by underlying project. I could override but don't want to. Alternatively, if you run `pnpm upgrade` and then `pnpm projen` those dep upgrades should not revert even though I thought they did before. Maybe something else changed... not 100% sure but it's working now.

**Comment by mrgrain** [comment, marked as answer]
Thanks for the question! Yes this is expected - to some extent. 

1. It is definitely the expectation that `projen default` is run __before__  code is merged into `main`. I usually do this locally. But if you have mutable builds enabled you can let the PR do the work for you (although when lockfiles are involved it is more complicated and often doesn't work). 
2. projen manages every dependency in one of two ways. This is not super well documented at the moment and not easy to check which approach is used for which dep.
    1. A **projen-managed dependency** will only ever use the dependency constraint that your projen app has set. You can try overwriting this with `project.addDeps('dep@version')`. But keep in mind that the component introducing it might have a good reason to declare a specific version constraint and things could start to break if you change it.
    2.  A **projen-declared** dependency allows you to manage the version inside `package.json` yourself. The difference is that in your projen app, it is only declared that "this dependency is needed" but no further constraint is specified. 

If there is a specific dependency that is annoying you, please let me know. It might be that we can lift the strict constraint in this case.
